private/bufpkg/bufimage/build_image.go

```diff
@@ -427,25 +426,132 @@ func newBuildImageOptions() *buildImageOptions {
 	return &buildImageOptions{}
 }
 
-// resolverForBuildResult implements protoencoding.Resolver and is backed
-// by a linker.Resolver which is the result of a protocompile operation.
-// The linker.Resolver provides all necessary methods except FindEnumByName.
-type resolverForBuildResult struct {
-	linker.Resolver
+// resolverForFiles implements protoencoding.Resolver and is backed
+// by a linker.Files and the *linker.Symbols symbol table produced
+// when compiling the files. The symbol table is used as an index
+// for more efficient lookup.
+type resolverForFiles struct {
+	pathToFile map[string]linker.File
+	symbols    *linker.Symbols
+}
```

Comment:
This is the meat of the change. This new resolver implementation is subtantially more efficient than the previous implementation (which leaned on a fairly naive implementation in protocompile's `linker` package).